File: Class_member/4.py
class Cal(object):
    _history = []
    def __init__(self,v1,v2):
        self.v1 = v1
        self.v2 = v2
    @classmethod
    def history(cls):
        for item in Cal._history:
            print(item)
    def add(self):
        result = self.v1 + self.v2
        Cal._history.append("add : %d + %d = %d" % (self.v1, self.v2, result))
        # Cal._history.append("add : " + str(result) )
        return result
    def subtract(self):
        result = self.v1 + self.v2
        Cal._history.append("subtract : %d - %d = %d" % (self.v1, self.v2, result))
        return result
    def multiply(self):
        result = self.v1 * self.v2
        Cal._history.append("multiply : %d * %d = %d" % (self.v1, self.v2, result))
        return result
    def divide(self):
        result = self.v1 / self.v2
        Cal._history.append("divide : %d / %d = %d" % (self.v1, self.v2, result))
        return result

c1 = Cal(10,20)
print(c1.add())
print(c1.subtract())
print(c1.multiply())
print(c1.divide())
print(Cal.history())

File: Class_member/4.rb
class Cal
  # attr_reader:v1, :v2
  # attr_writer:v1, :v2
  @@_history = []
  def initialize(v1, v2)
    @v1 = v1
    @v2 = v2
  end
  def add()
    result = @v1 + @v2
    # @@_history.push("add" + result.to_s())
    @@_history.push("add #{@v1} + #{@v2} = #{result}")
  end
  def subtract()
    result = @v1 - @v2
    @@_history.push("subtract #{@v1} - #{@v2} = #{result}")
  end
  def Cal.history()
    for item in @@_history
      p item
    end
  end
end

class CalMultiply < Cal
  def multiply()
    result = @v1 * @v2
    @@_history.push("multiply #{@v1} * #{@v2} = #{result}")
  end
end

class CalDivide < CalMultiply
  def divide()
    result = @v1 / @v2
    @@_history.push("divide #{@v1} / #{@v2} = #{result}")
  end
end

#
# c0 = Cal.new(10,20)
# p c0.add
#
# c1 = CalMultiply.new(10,20)
# # p c1, c1.multiply
# p c1.add
# # p c1.multiply

c2 = CalDivide.new(20,10)
# p c2.multiply
p c2.add
p c2.subtract
p c2.multiply
p c2.divide

Cal.history()
